File: backend/app/schemas/achievement.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List


class AchievementInfo(BaseModel):
    id: int
    code: str
    name: str
    description: str
    icon: str
    rarity: str
    condition_type: str
    condition_value: int
    exp_reward: int
    points_reward: int

    class Config:
        from_attributes = True


class AchievementCategory(BaseModel):
    category: str
    category_name: str
    achievements: List[AchievementInfo]


class AchievementListResponse(BaseModel):
    categories: List[AchievementCategory]


class UserAchievementItem(BaseModel):
    id: int
    code: str
    name: str
    description: str
    icon: str
    rarity: str
    category: str
    exp_reward: int
    points_reward: int
    unlocked_at: datetime


class UserAchievementResponse(BaseModel):
    unlocked: List[UserAchievementItem]
    newly_unlocked: List[UserAchievementItem] = []
    total_unlocked: int
    total_achievements: int
    completion_rate: float
    current_level: int
    current_exp: int
    next_level_exp: Optional[int]
    total_exp_rewarded: int
    total_points_rewarded: int


class AchievementProgressItem(BaseModel):
    achievement_id: int
    code: str
    name: str
    description: str
    icon: str
    rarity: str
    category: str
    condition_type: str
    exp_reward: int
    points_reward: int
    current_value: int
    target_value: int
    percentage: float
    is_unlocked: bool


class AchievementProgressResponse(BaseModel):
    progress: List[AchievementProgressItem]
    newly_unlocked: List[UserAchievementItem] = []

File: frontend/src/api/achievement.ts
import request from '@/utils/request'

export interface AchievementInfo {
  id: number
  code: string
  name: string
  description: string
  icon: string
  rarity: string
  condition_type: string
  condition_value: number
  exp_reward: number
  points_reward: number
}

export interface AchievementCategory {
  category: string
  category_name: string
  achievements: AchievementInfo[]
}

export interface UserAchievementItem {
  id: number
  code: string
  name: string
  description: string
  icon: string
  rarity: string
  category: string
  exp_reward: number
  points_reward: number
  unlocked_at: string
}

export interface AchievementProgressItem {
  achievement_id: number
  code: string
  name: string
  description: string
  icon: string
  rarity: string
  category: string
  condition_type: string
  exp_reward: number
  points_reward: number
  current_value: number
  target_value: number
  percentage: number
  is_unlocked: boolean
}

export interface AchievementMineResponse {
  unlocked: UserAchievementItem[]
  newly_unlocked: UserAchievementItem[]
  total_unlocked: number
  total_achievements: number
  completion_rate: number
  current_level: number
  current_exp: number
  next_level_exp: number | null
  total_exp_rewarded: number
  total_points_rewarded: number
}

export interface AchievementProgressResponse {
  progress: AchievementProgressItem[]
  newly_unlocked: UserAchievementItem[]
}

export const achievementApi = {
  async getAllAchievements(): Promise<{ categories: AchievementCategory[] }> {
    const res: any = await request.get('/achievements')
    return res.data || res
  },

  async getMyAchievements(): Promise<AchievementMineResponse> {
    const res: any = await request.get('/achievements/mine')
    return res.data || res
  },

  async getProgress(): Promise<AchievementProgressResponse> {
    const res: any = await request.get('/achievements/progress')
    return res.data || res
  }
}
